pkg/pillar/cmd/zedrouter/networkinstance.go

```diff
@@ -1870,3 +1860,178 @@ func uplinkToPhysdev(ifname string) string {
+	}
+	setFlag := false
+	for i := 0; i < len(subnet.IP); i++ {
+		if subnet.IP[i] != 0 {
+			setFlag = true
+			break
+		}
+	}
+	if setFlag == false {
+		return false
+	}
+	setFlag = false
+	for i := 0; i < len(subnet.Mask); i++ {
+		if subnet.Mask[i] != 0 {
+			setFlag = true
+			break
+		}
+	}
+	if setFlag == false {
+		return false
+	}
+	netStr := subnet.String()
+	if netStr == "" {
+		return false
+	}
+	if _, _, err := net.ParseCIDR(netStr); err != nil {
+		return false
+	}
+	return true
+}
+
+func getIPSubnetAddrCount(subnet net.IPNet) int {
+	networkStr := subnet.String()
+	maskLenStr := strings.Split(networkStr, "/")[1]
+	maskLen, _ := strconv.Atoi(maskLenStr)
+	addrcount := 1 << (32 - maskLen)
+	// discount the network, gateway and broadcast addresses(3)
+	addrcount -= 3
+	return addrcount
+}
+
+// do a basic validation for the config
+func isValidIPConfig(ctx *zedrouterContext,
+	status *types.NetworkInstanceStatus) bool {
+	if !isValidIPRange(status.DhcpRange) {
+		return isValidIPSubnet(status.Subnet)
+	}
+	return false
+}
+
+func validateIPConfig(ctx *zedrouterContext,
+	status *types.NetworkInstanceStatus) {
+	status.IsValidIPRange = isValidIPRange(status.DhcpRange)
+	status.IsValidIPSubnet = isValidIPSubnet(status.Subnet)
+	status.IsValidIPConfig = status.IsValidIPSubnet || status.IsValidIPRange
+	publishNetworkInstanceStatus(ctx, status)
+}
+
+// initialize the Host IP Address Management Module
+func handleNetworkInstanceIPAddrInit(ctx *zedrouterContext,
+	status *types.NetworkInstanceStatus) {
+
+	validateIPConfig(ctx, status)
+	if !status.IsValidIPConfig {
+		return
+	}
+	addrcount := 253
+	// get host IP Address count:
+	// if DhcpRange is set, derive from it
+	// otherwise derive from the subnet
+	if status.IsValidIPRange {
+		addrcount = getIPRangeAddrCount(status.DhcpRange)
+	} else {
+		if !status.IsValidIPSubnet {
+			return
+		}
+		addrcount = getIPSubnetAddrCount(status.Subnet)
+	}
+	// XXX: upper limit
+	if addrcount > 256 {
+		addrcount = 253
+	}
+	status.HostIPAdressCount = addrcount
+	log.Functionf("Initialized Network instance %s with addressable count %d\n",
+		status.DisplayName, addrcount)
+	publishNetworkInstanceStatus(ctx, status)
+}
+
+// for an App Inastance attached to an underlay network,
+// based on the appNum, try to get a free index within a
+// assignable ip address range
+func appNumOnUNetAllocate(ctx *zedrouterContext,
+	status *types.NetworkInstanceStatus, appNum int) int {
+	if !status.IsValidIPConfig {
+		return 0
+	}
+	// take the mod value for the range, on AppNum
+	appNumMod := appNum % status.HostIPAdressCount
+	appNumOnUNet := appNumMod
+
+	// if collision, find the next free bit
+	if status.AppNumBitMap.IsSet(appNumOnUNet) {
+		log.Functionf("AppNumOnUNetAllocate: Hash Collision detected for appNum %d on Network %s at %d\n",
+			appNum, status.DisplayName, appNumMod)
+		i := appNumMod + 1
+		// take next free bit value
+		for i = appNumMod + 1; i <= status.HostIPAdressCount; i++ {
```

Comment:
Wait, this doesn't provide a stable IP address across reboots when there is a collision. Hence the appNumOnNet only provides a stable IP address when apNumOnUNet is identical to appNum!

Since the appNum needs to be less than 256 already, and we already persist that number, it seems a lot simpler to 1) require that the appNum be less that the available IP addresses (253, or less if there is a DHCP range) and return an error otherwise 2) directly use the appNum as the IP address (adding to the base, which is either the start of the DHCP range or 1/2 in the subnet).

Reply:
For an IP address range more than or, equal to the Bitmap size, we will not have any collision. We mayhave a collision, for a very small DhcpRange and appNum crosses this count. And we most probably can not restrict the appNum, as it is global, and there can be many DhcpRange, across multiple network instances of variable size. Still need to workout a scheme to avoid the hash collision. Changed the code to use appNum as appNumOnUNet, for appNum smaller than HostIPAddressCount. For appNum more than, the HostIPAddressCount, there is still a risk for collision. I think, an handshake between the cloud and the EVE will solve this issue, wherei cloud always pushes a IP Address for the App Instance. And it would persist, with very simple implementation on EVE. Not sure. Please suggest.